src/Microsoft.AspNet.Http.Extensions/HeaderDictionaryTypeExtensions.cs

```diff
@@ -0,0 +1,202 @@
+            { typeof(StringWithQualityHeaderValue), new Func<IList<string>, IList<StringWithQualityHeaderValue>>(value => { IList<StringWithQualityHeaderValue> result; return StringWithQualityHeaderValue.TryParseList(value, out result) ? result : null; })  },
+            { typeof(CookieHeaderValue), new Func<IList<string>, IList<CookieHeaderValue>>(value => { IList<CookieHeaderValue> result; return CookieHeaderValue.TryParseList(value, out result) ? result : null; })  },
+            { typeof(EntityTagHeaderValue), new Func<IList<string>, IList<EntityTagHeaderValue>>(value => { IList<EntityTagHeaderValue> result; return EntityTagHeaderValue.TryParseList(value, out result) ? result : null; })  },
+            { typeof(SetCookieHeaderValue), new Func<IList<string>, IList<SetCookieHeaderValue>>(value => { IList<SetCookieHeaderValue> result; return SetCookieHeaderValue.TryParseList(value, out result) ? result : null; })  },
+        };
+
+        public static T Get<T>([NotNull] this IHeaderDictionary headers, string name)
+        {
+            object temp;
+            if (KnownParsers.TryGetValue(typeof(T), out temp))
+            {
+                var func = (Func<string, T>)temp;
+                return func(headers[name]);
+            }
+
+            var value = headers[name];
+            if (string.IsNullOrWhiteSpace(value))
+            {
+                return default(T);
+            }
+
+            return GetViaReflection<T>(value);
+        }
+
+        public static IList<T> GetList<T>([NotNull] this IHeaderDictionary headers, string name)
+        {
+            object temp;
+            if (KnownListParsers.TryGetValue(typeof(T), out temp))
+            {
+                var func = (Func<IList<string>, IList<T>>)temp;
+                return func(headers.GetValues(name));
+            }
+
+            var values = headers.GetValues(name);
+            if (values == null || !values.Any())
+            {
+                return null;
+            }
+
+            return GetListViaReflection<T>(values);
+        }
+
+        private static T GetViaReflection<T>(string value)
+        {
+            // TODO: Cache the reflected type for later? Only if success?
+            var type = typeof(T);
+            var method = type.GetMethods(BindingFlags.Public | BindingFlags.Static)
+                .Where(methodInfo =>
+                {
+                    if (string.Equals("TryParse", methodInfo.Name, StringComparison.Ordinal)
+                        && methodInfo.ReturnParameter.ParameterType.Equals(typeof(bool)))
+                    {
+                        var methodParams = methodInfo.GetParameters();
+                        return methodParams.Length == 2
+                            && methodParams[0].ParameterType.Equals(typeof(string))
+                            && methodParams[1].IsOut
+                            && methodParams[1].ParameterType.Equals(type.MakeByRefType());
```

Comment:
Can create the `ref` type just once outside of the LINQ query. Could make the code a smidgen clearer.